src/constructor/constructor.py

```diff
@@ -0,0 +1,259 @@
+        if not isinstance(parameters, Iterable) and not isinstance(parameters, Module) and \
+                not isinstance(parameters, Tensor):
+            raise ValueError(f'Unsupported parameters type for optimizer: {type(parameters)}')
+        elif not isinstance(parameters, Iterable):
+            parameters = [parameters]
+
+        param_groups = []
+        for model in parameters:
+            if isinstance(model, Module):
+                param_groups.extend(Constructor.__param_groups_weight_decay(model))
+            elif isinstance(model, Tensor) and model.requires_grad:
+                param_groups.append({'params': model})
+
+        return param_groups
+
+    # Adapted from https://github.com/rwightman/pytorch-image-models/blob/master/timm/optim/optim_factory.py
+    # Copyright 2019 Ross Wightman
+    # Licensed under The Apache 2.0 License [see LICENSE for details]
+    @staticmethod
+    def __param_groups_weight_decay(model: Module):
+        no_weight_decay_list = []
+        if hasattr(model, 'no_weight_decay'):
+            no_weight_decay_list = model.no_weight_decay()
+
+        decay = []
+        no_decay = []
+        for name, param in model.named_parameters():
+            if not param.requires_grad:
+                continue
+
+            if param.ndim <= 1 or name.endswith('.bias') or name in no_weight_decay_list:
+                no_decay.append(param)
+            else:
+                decay.append(param)
+
+        return [
+            {'params': no_decay, 'weight_decay': 0.},
+            {'params': decay}]
+
+    def create_dataloaders(self, phase: str) -> List[DataLoader]:
+        """Create data loaders.
+
+        Each data loader is based on a dataset while dataset consists
+        augmentations and transformations in the `albumentations`_ format.
+
+        Args:
+            phase: Phase for which the data loaders are to be built.
+            Should be one of: 'train', 'valid', 'test', 'predict'
+
+        Returns:
+            List of data loaders to be used inside `PyTorch Lightning Module`_
+        .. _albumentations: https://albumentations.ai/docs
+        .. _PyTorch Lightning Module:
+        https://pytorch-lightning.readthedocs.io/en/1.6.2/common/lightning_module.html#train-dataloader
+
+        Raises:
+            - ValueError: when requested phase is not from the specified list of supported phases
+            - ValueError: when transforms are not specified for composition augmentations of albumentation
+            - ValueError: when OneOrOther composition is passed that isn't supported
+        """
+        if phase not in ['train', 'valid', 'test', 'predict']:
+            raise ValueError(f'Not support phase for data loaders specification: {phase}')
+
+        return [
+            self.__prepare_dataloader(phase_params.dataset, phase_params.dataloader)
+            for phase_params in self.hparams.data['phase']
```

Comment:
Now config for dataloaders is 
`@dataclass
class DataloaderParams:
    # I think it must be list, with Enum Phase inside DataParams
    train: Optional[DataParams]
    valid: Optional[DataParams]
    test: Optional[DataParams]
    predict: Optional[DataParams]`
    I think we need set Phase inside DataParams and it would be Enum inside DataParams.
    Moreover now your code is not work, because self.hparams.data no key with name phase, you should use
    `[
            self.__prepare_dataloader(phase_params.dataset, phase_params.dataloader)
            for phase, phase_params in self.hparams.data.items()
            `